src/main/java/org/broadinstitute/dsm/jobs/TestBostonUPSTrackingJob.java

```diff
@@ -1,202 +1,211 @@
+    public void updateKitStatus(PoolingDataSource<PoolableConnection> dataSource, DdpKit kit, boolean isReturn, String ddpInstanceId, Config cfg) throws IOException {
         String trackingId;
         if (!isReturn) {
             trackingId = kit.getTrackingToId();
         }
         else {
             trackingId = kit.getTrackingReturnId();
         }
-
-        logger.info("Checking UPS status for " + trackingId + " for kit w/ external order number "+ kit.getExternalOrderNumber());
-        UPSTrackingResponse response = lookupTrackingInfo(trackingId);
-        logger.info("UPS response for " + trackingId + " is " + response);
-        String type;
+        String type = null;
+        UPSActivity lastActivity = null;
         if (isReturn) {
-            type = kit.getUpsReturnStatus();
+            if (kit.getKitReturnHistory() != null) {
+                lastActivity = new Gson().fromJson(kit.getKitReturnHistory(), UPSActivity[].class)[0];
+            }
         }
         else {
-            type = kit.getUpsTrackingStatus();
+            if (kit.getKitShippingHistory() != null) {
+                lastActivity = new Gson().fromJson(kit.getKitShippingHistory(), UPSActivity[].class)[0];
+            }
         }
-        if (StringUtils.isNotBlank(type)) {// get only type from it
-            type = type.substring(0, type.indexOf(' '));
+        if (lastActivity != null && lastActivity.getStatus().isDelivery()) {
+            this.logger.info("Tracking id " + trackingId + " is already delivered, not going to check UPS anymore");
+            return;
         }
+
+        logger.info("Checking UPS status for " + trackingId + " for kit w/ external order number " + kit.getExternalOrderNumber());
+        UPSTrackingResponse response = lookupTrackingInfo(cfg, trackingId);
+        logger.info("UPS response for " + trackingId + " is " + response);
+
         if (response != null && response.getErrors() == null) {
-            updateStatus(conn, trackingId, type, response, isReturn, kit, ddpInstance);
+            updateStatus(dataSource, trackingId, lastActivity, response, isReturn, kit, ddpInstanceId, cfg);
         }
         else {
             logError(trackingId, response.getErrors());
         }
     }
 
-    private static void logError(String trackingId, UPSError[] errors) {
+    private void logError(String trackingId, UPSError[] errors) {
         String errorString = "";
         for (UPSError error : errors) {
             errorString += "Got Error: " + error.getCode() + " " + error.getMessage() + " For Tracking Number " + trackingId;
         }
         logger.error(errorString);
     }
 
-    private static void updateStatus(@NonNull Connection conn, String trackingId, String oldType, UPSTrackingResponse response, boolean isReturn, DdpKit kit, DDPInstance ddpInstance) {
+
+    private void updateStatus(PoolingDataSource<PoolableConnection> dataSource, String trackingId, UPSActivity lastActivity, UPSTrackingResponse response,
```

Comment:
This should all be transactional, right?  If so, passing around the `dataSource` will potentially make this non-transactional, because each call to `getConnection()` can return a *different connection*, which may result in very weird bugs.  Please use a single connection instead of passing through the `dataSource`.